### PR TITLE
Implement sorting functionality and add total_supply for Universe Ass…

### PR DESCRIPTION
**Pull request by ben2077**
Implement sorting functionality and add total_supply for Universe Asset Stats

- Introduced ascending and descending sorting options for the Universe Asset Stats query.
- Added 'total_supply' field to Universe Asset Stats, enhancing the available dataset.
- Enhanced TestUniverseQuerySyncStatsSortingDirection to validate new sorting mechanisms and the inclusion of total_supply.
- Optimized SQL for efficient sorting based on user-selected parameters.

### CODE REVIEW

File: tapdb/sqlc/queries/universe.sql [hunk 1]
```diff
@@ -209,30 +209,20 @@ FROM asset_info
...
-        END,
-    CASE
-        WHEN sqlc.narg('sort_by') = 'total_proofs' THEN universe_stats.total_asset_proofs
-        ELSE NULL
-    END,
-    CASE
-        WHEN sqlc.narg('sort_by') = 'genesis_height' THEN asset_info.genesis_height
-        ELSE NULL
-    END
+    CASE WHEN sqlc.narg('sort_by') = 'asset_id' AND sqlc.narg('sort_direction') = 0 THEN   asset_info.asset_id END ASC,
```
Comment (guggero) [inline tapdb/sqlc/queries/universe.sql:212]: Can we unify the formatting here a bit so it's easier to read on narrow terminals? We aim at ~80 characters for those who like to have many rows of vim open.
Also, the ASC is kind of superfluous as it's the default.

My suggestion:

```
ORDER BY
    CASE WHEN sqlc.narg('sort_by') = 'asset_id' AND sqlc.narg('sort_direction') = 0 THEN
        asset_info.asset_id END,
    CASE WHEN sqlc.narg('sort_by') = 'asset_id' AND sqlc.narg('sort_direction') = 1 THEN
        asset_info.asset_id END DESC ,
    CASE WHEN sqlc.narg('sort_by') = 'asset_name' AND sqlc.narg('sort_direction') = 0 THEN
        asset_info.asset_name END,
    CASE WHEN sqlc.narg('sort_by') = 'asset_name' AND sqlc.narg('sort_direction') = 1 THEN
        asset_info.asset_name END DESC ,
    CASE WHEN sqlc.narg('sort_by') = 'asset_type' AND sqlc.narg('sort_direction') = 0 THEN
        asset_info.asset_type END,
    CASE WHEN sqlc.narg('sort_by') = 'asset_type' AND sqlc.narg('sort_direction') = 1 THEN
        asset_info.asset_type END DESC,
    CASE WHEN sqlc.narg('sort_by') = 'total_syncs' AND sqlc.narg('sort_direction') = 0 THEN
        universe_stats.total_asset_syncs END,
    CASE WHEN sqlc.narg('sort_by') = 'total_syncs' AND sqlc.narg('sort_direction') = 1 THEN
        universe_stats.total_asset_syncs END DESC,
    CASE WHEN sqlc.narg('sort_by') = 'total_proofs' AND sqlc.narg('sort_direction') = 0 THEN
        universe_stats.total_asset_proofs END,
    CASE WHEN sqlc.narg('sort_by') = 'total_proofs' AND sqlc.narg('sort_direction') = 1 THEN
        universe_stats.total_asset_proofs END DESC,
    CASE WHEN sqlc.narg('sort_by') = 'genesis_height' AND sqlc.narg('sort_direction') = 0 THEN
        asset_info.genesis_height END,
    CASE WHEN sqlc.narg('sort_by') = 'genesis_height' AND sqlc.narg('sort_direction') = 1 THEN
        asset_info.genesis_height END DESC,
    CASE WHEN sqlc.narg('sort_by') = 'total_supply' AND sqlc.narg('sort_direction') = 0 THEN
        asset_supply END,
    CASE WHEN sqlc.narg('sort_by') = 'total_supply' AND sqlc.narg('sort_direction') = 1 THEN
        asset_supply END DESC
LIMIT @num_limit OFFSET @num_offset;
```

File: tapdb/universe_stats_test.go [hunk 2]
```diff
@@ -247,81 +247,247 @@ func TestUniverseQuerySyncStatsSorting(t *testing.T) {
 
 	// sortCheck is used to generate an IsSorted func bound to the
 	// response, for each sort type below.
-	type sortCheck func([]universe.AssetSyncSnapshot) func(i, j int) bool
+	type sortCheck func([]universe.AssetSyncSnapshot, universe.SortDirection) func(i, j int) bool
```
Comment (guggero) [inline tapdb/universe_stats_test.go:250]: nit: line length, break before `universe.SortDirection`.

File: tapdb/universe_stats_test.go [hunk 3]
```diff
@@ -247,81 +247,247 @@ func TestUniverseQuerySyncStatsSorting(t *testing.T) {
...
+			name:      "asset name sort ascending",
+			sortType:  universe.SortByAssetName,
+			direction: universe.SortAscending,
 			isSortedFunc: func(s []universe.AssetSyncSnapshot,
-			) func(i, j int) bool {
+				d universe.SortDirection) func(i, j int) bool {
 
 				return func(i, j int) bool {
-					return s[i].AssetName < s[j].AssetName
+					if d == universe.SortAscending {
```
Comment (guggero) [inline tapdb/universe_stats_test.go:268]: We could extract all of these individual `isSortedFunc` into a single one, now that they have become a bit more complicated.
For example:
```
	isSorted := func(s []universe.AssetSyncSnapshot,
		t universe.SyncStatsSort,
		d universe.SortDirection) func(i, j int) bool {

		asc := d == universe.SortAscending
		desc := d == universe.SortDescending
			
		return func(i, j int) bool {
			switch {
			case t == universe.SortByAssetName && asc:
				return s[i].AssetName < s[j].AssetName
			case t == universe.SortByAssetName && desc:
				return s[i].AssetName > s[j].AssetName
				
				...
			}
		}
	}
```

File: tapdb/universe_stats_test.go [hunk 4]
```diff
@@ -247,81 +247,247 @@ func TestUniverseQuerySyncStatsSorting(t *testing.T) {
...
+							s[j].GenesisHeight
+					}
+				}
+			},
+		},
+		{
+			name:      "total supply descending",
+			sortType:  universe.SortByTotalSupply,
+			direction: universe.SortDescending,
+			isSortedFunc: func(s []universe.AssetSyncSnapshot, d universe.SortDirection) func(i, j int) bool {
```
Comment (guggero) [inline tapdb/universe_stats_test.go:472]: nit: line length/formatting here and below.

File: tapdb/universe_stats_test.go [hunk 5]
```diff
@@ -330,15 +496,19 @@ func TestUniverseQuerySyncStatsSorting(t *testing.T) {
...
 			require.NoError(t, err)
 			require.Len(t, syncStats.SyncStats, numAssets)
+			for _, assetStat := range syncStats.SyncStats {
+				println(assetStat.TotalSupply)
+			}
 
 			require.True(t, sort.SliceIsSorted(
 				syncStats.SyncStats,
-				testCase.isSortedFunc(syncStats.SyncStats),
+				testCase.isSortedFunc(syncStats.SyncStats, testCase.direction),
```
Comment (guggero) [inline tapdb/universe_stats_test.go:511]: nit: line length

File: universe/interface.go [hunk 6]
```diff
@@ -531,6 +531,18 @@ const (
...
+	SortByTotalSupply
+)
+
+type SortDirection uint8
+
+const (
+	// SortAscending is a sentinel value that indicates that the sort
+	// should be in ascending order.
+	SortAscending SortDirection = iota
+	SortDescending
```
Comment (guggero) [inline universe/interface.go:545]: nit: type definition and constant is missing Godoc.

File: universe/interface.go [hunk 7]
```diff
@@ -551,6 +563,10 @@ type SyncStatsQuery struct {
 	// SortBy is the sort order to use when returning the stats.
 	SortBy SyncStatsSort
 
+	// SortDesc is a boolean that indicates if the sort should be in
```
Comment (guggero) [inline universe/interface.go:566]: nit: Godoc name doesn't match member name.

**Review by guggero — COMMENTED** [review]
Thanks a lot for the PR!
Looks pretty good, main comments are around code style/optimizations.

**Comment by dstadulis** [comment]
Thanks very much for the PR! @jamaljsr will test shortly

**Comment by jamaljsr** [comment]
tACK. I tested the functionality. LGTM 👌

File: tapdb/universe_stats_test.go [hunk 8]
```diff
@@ -247,98 +247,231 @@ func TestUniverseQuerySyncStatsSorting(t *testing.T) {
...
+			sortType:  universe.SortByAssetName,
+			direction: universe.SortAscending,
 			isSortedFunc: func(s []universe.AssetSyncSnapshot,
-			) func(i, j int) bool {
+				d universe.SortDirection) func(i, j int) bool {
 
-				return func(i, j int) bool {
-					return s[i].AssetName < s[j].AssetName
-				}
+				return isSortedWithDirection(s,
```
Comment (guggero) [inline tapdb/universe_stats_test.go:312]: nit: we format multi-line function calls like this:
```
				return isSortedWithDirection(
					s, universe.SortByAssetName, d,
				)
```

File: universe/interface.go [hunk 9]
```diff
@@ -531,6 +531,22 @@ const (
 
 	// SortByGenesisHeight sorts the returned stats by the genesis height.
 	SortByGenesisHeight
+
+	// SortByTotalSupply sorts the returned stats by the total supply.
+	SortByTotalSupply
+)
+
+// SortDirection is an enum used to specify the sort direction of the returned
```
Comment (guggero) [inline universe/interface.go:539]: nit: missing full stop.

**Review by guggero — APPROVED** [review]
A few more nits, then this is good to go.
And can you please squash your commits into a single one? Thanks!

**Review by Roasbeef — APPROVED** [review]
LGTM 🪁

Thanks for such a great first contribution PR!

**Pull request merged**